snuba/subscriptions/data.py

```diff
@@ -75,6 +75,7 @@ class SubscriptionData:
     """
 
     project_id: int
+    organization: Optional[int]
```

Comment:
This approach would have an issue. Not all datasets know about organizations. Errors and transactions do not have an organization column. 
We did not put the organization id here back when we introduced sessions subscriptions exactly because it was not a common field to all datasets types we wanted to run subscriptions onto. 
Thus the SubscriptionEntity. 

This seems to go against a design where subscriptions are customizable per dataset (they are exactly because not everybody has the same fields and validators).

Reply:
If your goal is to make `SubscriptionEntity` immutable and static, which can be a good idea, why not just putting the logic into`EntitySubscription` ? 
What we needed `EntitySubscription` for was to process the subscription query before executing in a way that was specific to the `Entity`. There is nothing `EntitySubscription` needs that is not already in the query: like the condition on the organization. 
This is how the EntitySubscription is used 
https://github.com/getsentry/snuba/blob/master/snuba/subscriptions/data.py#L145-L169
The entity subscription logic could be executed after the parsing and could extract the query condition from the query itself.

Reply:
I don't think we are breaking that design necessarily. The organization data was already being stored in redis alongside all these other fields. Instead of reading that field out and storing it in a different class from the other fields, we are simply putting into this class.

The customization is still possible, because the entity can configure whether or not to use these fields.

Reply:
Wouldn't that mean that, if tomorrow profiles decides to add a new field to the subscription data class, we would add a profile specific field in the generic `SubscriptionData` ? 

Though, taking a step back, I am not 100% sure this whole problem needs to exist. 
@lynnagara  do you remember why we needed to have organization id as a separate field in the SubscriptionData representation in redis? That is constant for a subscription id, it is not like the time window that changes at every iteration. Is there any reason the `organization_id` condition cannot just be in the SnQL query string ?

@evanh , @enochtangg you should also ensure any design change is in line with what the Slicing project is planning to change around subscriptions: https://www.notion.so/sentry/Generic-Metrics-Slicing-9c206e02f92a4abca5948d2e21c6c00d#cdb80a402b894440adf05cf1805363ec

Reply:
> Though, taking a step back, I am not 100% sure this whole problem needs to exist. @lynnagara do you remember why we needed to have organization id as a separate field in the SubscriptionData representation in redis? That is constant for a subscription id, it is not like the time window that changes at every iteration. Is there any reason the `organization_id` condition cannot just be in the SnQL query string ?
I believe it was done for the same reason that project_id is not in the SnQL query string. We need to be extra careful to ensure there is only one.

Reply:
> I believe it was done for the same reason that project_id is not in the SnQL query string. We need to be extra careful to ensure there is only one.

Project id, was not there because of historical reasons:
- it was before SnQL. project id was not in the query at that time.
- project id was supposed to be known by the pipeline for semantic partitioning, and before the AST there was no way. 

So in both cases, this should not have applied to organization id when this was introduced and if these are the reasons  it should not be needed anymore, unless it is needed somehow in the results.
I would consider getting rid of it and fixing the content of the storage.

Reply:
> it should not be needed anymore, unless it is needed somehow in the results

We would need to run a migration to update all the subscriptions in Redis to make sure all the ones with an `organization` tag also have an organization condition. This feels like a lot of unnecessary work. Even if we don't remove it, organization is a concept that is useful across many datasets.